### ISSUE TITLE
typed_triggers hook ruby block with nil machine argument

### ISSUE THREAD

**Issue by rgl** [hashicorp/vagrant#11513]
### Vagrant version

2.2.7

### Host operating system

Windows 2019

### Guest operating system

Ubuntu 18.04

### Vagrantfile

```ruby
  # NB this is a global trigger.
  # NB you can only add hook triggers globally.
  config.trigger.before 'before_start_instance', type: :hook do |trigger|
    trigger.only_on = 'default'
    trigger.ruby do |env, machine|
      raise 'ALWAYS RAISED' unless machine
    end
  end
```

### Workaround

This ugly workaround works:

```ruby
  # NB this is a global trigger.
  # NB you can only add hook triggers globally.
  config.trigger.before 'before_start_instance', type: :hook do |trigger|
    trigger.only_on = 'default'
    trigger.ruby do |env, machine|
      # NB because the trigger.ruby machine argument is always nil, we have to
      #    hack our way into finding the machine instance.
      env.active_machines.each do |machine_name, machine_provider|
        if machine_provider.to_s == 'hyperv' && machine_name.to_s == trigger.only_on
          m = env.machine(machine_name, machine_provider)
          # m is never nil here
        end
      end
    end
  end
```

**Comment by chrisroberts** [comment]
@rgl Hi and thanks for the report! The implementation details for how the triggers interact with the call stack have been updated very recently (#11455). Would you be able to give your Vagrantfile a run with a nightly build to see if it resolves the issue? Your trigger definition should also be able to be defined without the `before_` prefix (`config.trigger.before :start_instance, type: :hook do`).

Nightly builds are here: https://github.com/hashicorp/vagrant-installers/releases/tag/v2.2.8.dev%2Bmaster

Thanks!

**Comment by rgl** [comment]
Tried to new version, the before_start_instance (or start_instance) hook no longer exists. These are the hooks that vagrant tried to call:

```plain
INFO environment: Running hook: authenticate_box_url
INFO environment: Running hook: environment_plugins_loaded
INFO environment: Running hook: environment_load
```

What seems to exist now is a provider specific action like:

```plain
 INFO warden: Calling IN action: #<VagrantPlugins::HyperV::Action::StartInstance:0x000000000647bd68>
```

And using it works nicely in a ruby trigger block!

A little nit pick, there is an extra space before `...` in the output of vagrant while calling the action:

```plain
==> default: Running action triggers before VagrantPlugins::HyperV::Action::StartInstance ...
```

BTW, will vagrant 2.2.8 have typed triggers out of experimental?

**Comment by chrisroberts** [comment]
@rgl Thanks for giving the nightly a try! I'll get that extra space out of there too :slightly_smiling_face: 

The typed triggers will be moving out of experimental in the 2.3.0 release (along with some other things). Aside from the updates they have received in this upcoming release, we are ensuring use cases are covered and getting documentation updated where required. So it will be generally available soon, and thanks so much for testing and providing feedback!

Since this looks to be fixed by #11455 I'm going to close this up.

Cheers!
